### DISCUSSION TITLE
Auto-load settings file based on active aircraft

### DISCUSSION THREAD

**Discussion by neilenns** [MobiFlight/MobiFlight-Connector#419] [Ideas]
I maintain several different MobiFlight configurations that map to the different planes I fly in FS2020 (one covers the C172 G1000, the other the TBM930).

I have to remember to manually open the right file when flying to make sure the switches and whatnot are set up properly, and often forget until I'm starting to configure the plane in the SIM for takeoff.

Since MF now has SimConnect access it would be nice if I could tag configurations with different aircraft and have MobiFlight automatically load the right config when it detects an aircraft change.

(This is how AxisAndOhs works, and if there's plans long term for MobiFlight to support joystick inputs as well this would be a really great addition).

**Comment by JaimeLeon2** [comment]
I think we sort of have a list of events per airplane in the online spreadsheet. The problem is that there is a bunch of events that are obsolete so a cleanup effort is necessary. 

Second, the bridge to MF Connector is through events.txt, with just one level of groups which implies a large number of groups. It also requires you to know in which group to look for an event.

**Reply by neilenns** [reply]
Yes, that list is great for helping people set up their config without having to do a ton of digging. But it's a little different from auto-loading pre-defined configurations based on planes.

I have a config for G1000 and a config for G3000 aircraft since my mappings are different for those two different glass cockpits. Right now I have to remember to open the right settings file when I'm flying the C712 vs. the TBM930. It'd be cool if MobiFlight just detected the plane and loaded the configs for me automatically.

**Comment by MobiFlight-Admin** [comment]
Yeah, I understand. It would be definitely a cool convenience feature. 
From a priority point of view, I think with the recent file list being available, loading a different plane configuration is pretty easy.

I have an idea how to add it though, it should also work for P3D. We should not forget, that we also have a user base on other sims.

**Reply by FSXLiveriesdotcom** [reply]
If you will make it available, it will be great. If you will make it available for multiple sims, it will be mind blowing! :)

**Comment by pizman** [comment]
About older Sims ....
FSUIPC allow to "read" the current loaded Aircraft as a String .    Pretty sure via WASM/Simconnect the same Data is available in FS2020.

Whatever i see no real practicable way to do this.  A Automatic Load Routine will be confusing and difficult to setup..... A Manual Click Load finally remove the benefit.  If i need to click on a bar i could also click to LOAD and select the needed file.

For sure..... Not everybody is like myself or Stephan and only fly one single aircraft ( PMDG) at all time.    But is there a real usecase ?
Whatever if you fly 10 different Aircrafts.... You pretty sure will not change them every 5 minutes.
So lets say you fly 5 Hours at afternoon and yes lets say you use 4 different Aircrafts that day......  Is it really a mess to click on Load in MF every 60-90 Mins one time ?
The FS2020 need up to 2 Minutes to Load a Scenario ( Aircraft swap).     So enough time to shortly get MF from the Taskbar and load the file !

**Reply by neilenns** [reply]
Axis and Ohs has this feature and it's wonderful, that's why I opened the request :) It makes for one less thing to set up when you are trying to get ready to fly.

**Reply by FSXLiveriesdotcom** [reply]
Yeah, to be fair for people who fly several aircrafts, such of auto profile switch would be very appreciated. That's something what I always liked about FSUIPC and LINDA because I didn't have to think about. 
One thing less to think about =  better workflow aka shorter checklist when running all 3rd apps.

**Comment by Gamma36** [comment]
I think it is essential to load a config for each plane type, when the user presses Fly on FS2020 for example.

I know that the MCP commands for A320NX, 787 and 747 are different, likewise all the planes with advanced avionics have different commands. Juggling different configs becomes increasingly complex when building a cockpit for many planes. Likewise testing different configs becomes complex. 

Associating a config with a particular plane _and_ a particular airline (max flaps can be different) is much slicker than fumbling around a directory for the correct config file.

Further to that, anything that helps automate the propagation of Arduino hardware changes across the multiple configs is of value.

**Comment by JaimeLeon2** [comment]
msfs-popout-manager has this cool auto load feature.  The user specifies a profile to be linked to the airplane label obtained from the MSFS2020.  Once this link is registered. The app can find, move and resize all popout screens, for any airplane previously defined, with the user just pressing one button.  It is a very convenient feature.  I would be great if we could have this level of convenience in Mobiflight.

**Comment by FSH1HH** [comment]
I would love such a feature!

**Comment by Oudoum** [comment]
Loaded aircraft can be requested via SimConnect and the RequestSystemState function.
https://docs.flightsimulator.com/html/Programming_Tools/SimConnect/API_Reference/General/SimConnect_RequestSystemState.htm

But adding the logic will take several days.

**Reply by JaimeLeon2** [reply]
I think one current limitation for this project is that Mobiflight cannot read strings directly from Simconnect.
One has to go through Fsuipc.
I can read (A:TITLE, string) but only through fsuipc7.

**Reply by Oudoum** [reply]
Why should MobiFlight not be able to read strings via SimConnect?

**Reply by JaimeLeon2** [reply]
That I don't know. We need to ask Sebastian about that. I guess the wasm module does not know how to handle string type values.

**Reply by Oudoum** [reply]
You do not need to get the SystemState via the WASM module. But even if you want to do you will receive this struct:

```
typedef struct SIMCONNECT_RECV_SYSTEM_STATE : public SIMCONNECT_RECV {
    DWORD  dwRequestID;
    DWORD  dwInteger;
    float  fFloat;
    char   szString[MAX_PATH];
    };
```
    
Take the szString, but this much more trouble receiving from SimConnect in the WASM module and then send it back to SimConnect to receive it in a client.

**Reply by DocMoebiuz** [reply]
Aint working in MSFS2020 ... i tried but I solved it differently.

**Reply by Oudoum** [reply]
> Aint working in MSFS2020 ... i tried but I solved it differently.

I can show you how I did it in my code.
It should work because this has not changed since FSX.

**Reply by neilenns** [reply]
Support for this was released in the latest beta.

**Comment by Gamma36** [comment]
Aircraft.cfg is the file that is unique to each installed MSFS plane. 

It contains a lot of modern data, including ICAO designators for manufacturer (Boeing), type (B738). model (737-800), engine type (Jet) and engine count (2).

So in my opinion that is the data to go after, as it is very comprehensive and specific to the aircraft variant chosen by the user.

https://docs.flightsimulator.com/html/Content_Configuration/SimObjects/Aircraft_SimO/aircraft_cfg.htm
